platforms/native/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/plugins/ClangCompilerPlugin.java

```diff
@@ -48,28 +37,8 @@ public abstract class ClangCompilerPlugin implements Plugin<Project> {
     @Override
     public void apply(Project project) {
         project.getPluginManager().apply(NativeComponentPlugin.class);
-    }
-
-    static class Rules extends RuleSource {
-        @Defaults
-        public static void addToolChain(NativeToolChainRegistryInternal toolChainRegistry, ServiceRegistry serviceRegistry) {
-            final FileResolver fileResolver = serviceRegistry.get(FileResolver.class);
-            final ExecActionFactory execActionFactory = serviceRegistry.get(ExecActionFactory.class);
-            final CompilerOutputFileNamingSchemeFactory compilerOutputFileNamingSchemeFactory = serviceRegistry.get(CompilerOutputFileNamingSchemeFactory.class);
-            final Instantiator instantiator = serviceRegistry.get(Instantiator.class);
-            final BuildOperationExecutor buildOperationExecutor = serviceRegistry.get(BuildOperationExecutor.class);
-            final CompilerMetaDataProviderFactory metaDataProviderFactory = serviceRegistry.get(CompilerMetaDataProviderFactory.class);
-            final SystemLibraryDiscovery standardLibraryDiscovery = serviceRegistry.get(SystemLibraryDiscovery.class);
-            final WorkerLeaseService workerLeaseService = serviceRegistry.get(WorkerLeaseService.class);
-
-            toolChainRegistry.registerFactory(Clang.class, new NamedDomainObjectFactory<Clang>() {
-                @Override
-                public Clang create(String name) {
-                    return instantiator.newInstance(ClangToolChain.class, name, buildOperationExecutor, OperatingSystem.current(), fileResolver, execActionFactory, compilerOutputFileNamingSchemeFactory, metaDataProviderFactory, standardLibraryDiscovery, instantiator, workerLeaseService);
-                }
-            });
-            toolChainRegistry.registerDefaultToolChain(ClangToolChain.DEFAULT_NAME, Clang.class);
-        }
-
+        NativeToolChainRegistryInternal toolChainRegistry = Cast.uncheckedCast(project.getExtensions().getByType(NativeToolChainRegistry.class));
```

Comment:
💅 If we expose this as `NativeToolChainRegistryInternal` in `NativeBinaryServices`, we wouldn't have to cast here.

Reply:
I thought about this too, but it's nice to have the public API in the service registry and extension container.  Otherwise, the internal type leaks out more.